homeassistant/components/gree/__init__.py

```diff
@@ -0,0 +1,77 @@
+"""The Gree Climate integration."""
+import logging
+
+from homeassistant.components.climate import DOMAIN as CLIMATE_DOMAIN
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+
+from .bridge import CannotConnect, DeviceHelper
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the Gree Climate component."""
+    hass.data[DOMAIN] = {"devices": {}, "pending": {}}
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up Gree Climate from a config entry."""
+
+    device_infos = []
+    devices = []
+
+    # First we'll grab as many devices as we can find on the network
+    # it's necessary to bind static devices anyway
+    _LOGGER.debug("Scanning network for Gree devices...")
+
+    try:
+        for device_info in await DeviceHelper.find_devices():
+            device_infos.append(device_info)
+    except Exception as exception:  # pylint: disable=broad-except
```

Comment:
I think it has to be pretty broad here. This is a fairly thin adaptor around a UDP BROADCAST send/recv sequence. So the exception could be anything thrown by the underlying socket (which could have an OS specific impact) and it's not atypical for timeout possible other errors as the underlying system will test broadcast on any ipv4 interface it finds, and then any attempt to decode that packet into a device data structure.

I've already a fully async version of the code ready in another branch, waiting on this to be merged before I submit the next PR. It will check that status of the underlying tasks and discard errors at a lower level, I will need to double check but it should get rid of the need for exception handling at this level.

Reply:
If it's a socket we should use `OSError`.

Reply:
Errors from the data parsing should generally be trapped by the library since those concern details of the device / service output.

Preferably the library should catch all errors and raise library specific errors as needed.

Reply:
I forgot that I had already basically done this stuff :)